lib/Runtime/Library/JavascriptMap.cpp

```diff
@@ -6,462 +6,694 @@
+        }
+        // If the key isn't in the map, check if the canonical value is
+        Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(key);
+        // If the simple var is the same as the original key, we know it isn't in the map
+        if (!simpleVar || simpleVar == key)
+        {
+            return false;
         }
 
-        JavascriptMap* map = JavascriptMap::FromVar(args[0]);
-        return scriptContext->GetLibrary()->CreateMapIterator(map, JavascriptMapIteratorKind::Value);
-    }
-
-    void JavascriptMap::Clear()
-    {
-        list.Clear();
-        map->Clear();
+        if (!this->u.simpleVarMap->TryGetValue(simpleVar, &node))
+        {
+            return false;
+        }
+        *value = node->data.Value();
+        return true;
     }
-
-    bool JavascriptMap::Delete(Var key)
+    case MapKind::ComplexVarMap:
     {
-        if (map->ContainsKey(key))
+        MapDataNode * node = nullptr;
+        if (!this->u.complexVarMap->TryGetValue(key, &node))
         {
-            MapDataNode* node = map->Item(key);
-            list.Remove(node);
-            return map->Remove(key);
+            return false;
         }
+        *value = node->data.Value();
+        return true;
+    }
+    default:
+        Assume(UNREACHED);
         return false;
     }
+}
+
+
+bool
+JavascriptMap::Has(Var key)
+{
+    JS_REENTRANCY_LOCK(jsReentLock, this->GetScriptContext()->GetThreadContext());
 
-    bool JavascriptMap::Get(Var key, Var* value)
+    switch (this->kind)
     {
-        if (map->ContainsKey(key))
+    case MapKind::EmptyMap:
+        return false;
+
+    case MapKind::SimpleVarMap:
+    {
+        // First check if the key is in the map
+        if (this->u.simpleVarMap->ContainsKey(key))
         {
-            MapDataNode* node = map->Item(key);
-            *value = node->data.Value();
             return true;
         }
+        // If the key isn't in the map, check if the canonical value is
+        Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(key);
+        // If the simple var is the same as the original key, we know it isn't in the map
+        if (!simpleVar || simpleVar == key)
+        {
+            return false;
+        }
+
+        return this->u.simpleVarMap->ContainsKey(simpleVar);
+    }
+    case MapKind::ComplexVarMap:
+        return this->u.complexVarMap->ContainsKey(key);
+
+    default:
+        Assume(UNREACHED);
         return false;
     }
+}
+
+void
+JavascriptMap::PromoteToComplexVarMap()
+{
+    AssertOrFailFast(this->kind == MapKind::SimpleVarMap);
 
-    bool JavascriptMap::Has(Var key)
+    uint newMapSize = this->u.simpleVarMap->Count() + 1;
+    ComplexVarDataMap* newMap = RecyclerNew(this->GetRecycler(), ComplexVarDataMap, this->GetRecycler(), newMapSize);
+
+    JavascriptMap::MapDataList::Iterator iter = this->list.GetIterator();
+    // TODO: we can use a more efficient Iterator, since we know there will be no side effects
```

Comment:
On the topic of possible optimizations for this piece, we should be able to creepily swap out the vtable on simpleVarMap and just claim it's now a ComplexVarMap, as long as the hash functions match (note my other comment on using RecyclerPointerComparer, which would be a prerequisite). I think copying is fine because it's much clearer and we only do it once per Map or Set, but we might want to include this option as part of the TODO.